File: cpp/d5.cpp
#include "aoc.h"

int main() {
    ifstream file("puzzles/d5.txt");
    unordered_map<int, vector<int>> rules;
    string line;
    bool firstPortion = true;
    int ans1{}, ans2{};
    while (getline(file, line)) {
        if (line == "") {
            firstPortion = false;
            continue;
        }

        stringstream ss(line);
        char delimeter{};
        if (firstPortion) {
            int num1{}, num2{};
            ss >> num1 >> delimeter >> num2;
            rules[num1].push_back(num2);
        }
        else {
            vector<int> update;
            int num;
            while (ss >> num) {
                update.push_back(num);
                ss >> delimeter;
            }
            bool ok = true;
            for (int i=0; i<update.size()-1; i++) {
                for (int j=i+1; j<update.size(); j++) {
                    if (auto& curr = rules[update[j]]; find(curr, update[i]) != curr.end()) {
                        ok = false;
                        int value = update[j];
                        update.erase(update.begin()+j);
                        update.insert(update.begin()+i, value);
                    }
                }
            }
            if (ok) ans1 += update[update.size()/2];
            else ans2 += update[update.size()/2];
        }
    }
    cout << ans1 << "   " << ans2 << endl;
}

File: js/d5.js
const fs = require("fs");

fs.readFile("puzzles/d5.txt", "utf-8", (lol, data) => {
    const rules = {};
    const lines = data.split(/\n/);
    let firstPart = true;
    let ans1 = 0;
    let ans2 = 0;
    for (const line of lines) {
        if (line == "") firstPart = false;
        else if (firstPart) {
            const [a, b] = line.split(/\|/);
            if (!rules[a]) {
                rules[a] = [];
            }
            rules[a].push(b);
        } else {
            let ok = true;
            const currLine = line.split(/\,/);
            for (let i = 1; i < currLine.length; i++) {
                for (let j = 0; j < i; j++) {
                    if (rules[currLine[i]].includes(currLine[j])) {
                        ok = false;
                        [currLine[i], currLine[j]] = [currLine[j], currLine[i]];
                    }
                }
            }
            if (ok) {
                ans1 += +currLine[Math.floor(currLine.length / 2)];
            } else {
                ans2 += +currLine[Math.floor(currLine.length / 2)];
            }
        }
    }
    console.log("part1:", ans1, "part2:", ans2);
});
